userspace/libsinsp/container_info.cpp

```diff
@@ -150,7 +150,7 @@ const sinsp_container_info::container_mount_info *sinsp_container_info::mount_by
 
 std::shared_ptr<sinsp_threadinfo> sinsp_container_info::get_tinfo(sinsp* inspector) const
 {
-	std::shared_ptr<sinsp_threadinfo> tinfo(inspector->build_threadinfo());
+	std::shared_ptr<sinsp_threadinfo> tinfo(inspector->build_threadinfo().release());
```

Comment:
This seems sorta silly, why are we creating a unique pointer for it to just be released and then passed into a shared pointer? I would argue rolling back `new_threadinfo` to return a raw pointer is a better alternative, since it is only used in `build_threadinfo` and focus on making sure callers of `build_threadinfo` understand they are the owners of the returned pointer and should handle releasing the resource.

I understand raw pointers are not great, that doesn't mean we should mindlessly remove them, but rather understand where it makes sense to use them and limit their scope as much as possible. In this case, creating an object (the `std::unique_ptr`) only for it to be deleted seems like an unnecessary step.

Reply:
I generally don't agree with this view. Unique pointer serve both for managing the lifecycle of the owned pointer, but also for making it clear who the owner of that pointer is. `build_threadinfo` is a factory method, and thus is perfectly suitable for returning a unique pointer as it clearly communicates (and enforces) that the ownership of the pointer is of the caller and that the inspector will own nothing about it from that point onwards.

The operation of transforming a pointer from unique-ownership to shared-ownership is quite common and has sense, to the point in which `std::shared_ptr` also has a move constuctor for `std::unique_ptr`, which could be used in this snippet also as an alternative.

I would also like to underline that unlike `std::shared_ptr` (that do ref counting), `std::unique_ptr` has mostly no cost if not in the case in which it is destroyed while handling a pointer. Invoking `release` simply makes the unique pointer release ownership without destroying anything if not itself. My guess is that compilers are also smart enough to inline most of this.

The goal of this PR, and also of safer memory management in general, is to reduce human error to the bare minimum by not making anyone "guessing" what the lifecycle of a pointer is, but rather making it clear and compiler/runtime checked.

As for whether it makes sense to have a `std::shared_ptr` in this specific snippet or not -- it probably has not. As stated in the PR description, there are many places in the codebase in which we adopt `std::shared_ptr` just for the sake of it, and that definitely has a cost. I preferred not to tackle this problem in this contribution otherwise I would have made this PR too bloated, and this is also why sometime you see this unique->shared ptr conversions (which again should have no different cost compared to what we did before).